protocol/src/protocol.callHierarchy.proposed.ts

```diff
@@ -0,0 +1,172 @@
+ * is of type [CallHierarchyItem](#CallHierarchyItem) or a Thenable that resolves to such.
+ */
+export namespace CallHierarchyResolveRequest {
+    export const type = new RequestType<ResolveCallHierarchyItemParams, CallHierarchyItem, void, void>('callHierarchy/resolve');
+    export type HandlerSignature = RequestHandler<ResolveCallHierarchyItemParams, CallHierarchyItem | null, void>;
+}
+
+/**
+ * The parameters of a `textDocument/callHierarchy` request.
+ */
+export interface CallHierarchyParams extends TextDocumentPositionParams {
+    /**
+     * The number of levels to resolve.
+     */
+    resolve?: number;
+    /**
+     * The direction of calls to resolve.
+     */
+    direction?: CallHierarchyDirection;
+}
+
+/**
+ * The parameters of a `callHierarchy/resolve` request.
+ */
+export interface ResolveCallHierarchyItemParams {
+    /**
+     * Unresolved item.
+     */
+    item: CallHierarchyItem;
+    /**
+     * The number of levels to resolve.
+     */
+    resolve: number;
+    /**
+     * The direction of calls to resolve.
+     */
+    direction: CallHierarchyDirection;
+}
+
+/**
+ * The direction of a call hierarchy.
+ */
+export namespace CallHierarchyDirection {
+    /**
+     * The callers of a symbol.
+     */
+    export const Incoming: 1 = 1;
+
+    /**
+     * The callees of a symbol.
+     */
+    export const Outgoing: 2 = 2;
+}
+
+export type CallHierarchyDirection = 1 | 2;
+
+/**
+ * The result of a `textDocument/callHierarchy` request.
+ */
+export interface CallHierarchyItem {
+
+    /**
+     * The name of the symbol targeted by the call hierarchy request.
+     */
+    name: string;
+
+    /**
+     * More detail for this symbol, e.g the signature of a function.
+     */
+    detail?: string;
+
+    /**
+     * The kind of this symbol.
+     */
+    kind: SymbolKind;
+
+    /**
+     * `true` if the hierarchy item is deprecated. Otherwise, `false`. It is `false` by default.
+     */
+    deprecated?: boolean;
+
+    /**
+     * URI of the document containing the symbol.
+     */
+    uri: string;
+
+    /**
+     * The range enclosing this symbol not including leading/trailing whitespace but everything else
+     * like comments. This information is typically used to determine if the the clients cursor is
+     * inside the symbol to reveal in the symbol in the UI.
+     */
+    range: Range;
+
+    /**
+     * The range that should be selected and revealed when this symbol is being picked, e.g the name of a function.
+     * Must be contained by the the `range`.
+     */
+    selectionRange: Range;
+
+    /**
+     * The actual locations of incoming (or outgoing) calls to (or from) a callable identified by this item.
+     *
+     * *Note*: undefined in root item.
+     */
+    callLocations?: Location[];
+
+    /**
+     * List of incoming (or outgoing) calls to (or from) a callable identified by this item.
+     *
+     * *Note*: if undefined, this item is unresolved.
+     */
+    calls?: CallHierarchyItem[];
```

Comment:
What is the relationship between `calls` and `callLocations`? Does one location correspond to one call? E.g could this be though of as array of tuples, like `calls: [Location, CallHierarchyItem][]`?

Reply:
> E.g could this be though of as array of tuples, like calls: [Location, CallHierarchyItem][]

We even [discussed](https://github.com/Microsoft/vscode-languageserver-node/pull/420#issuecomment-458109531), that this should be `[Location[], CallHierarchyItem][]` to squash multiple calls to/from the the same callable item.


> Yeah, I am all in for explicit information but my focus is also the vscode API which is less about the transport of data.

My point _is_ about the API: if you separate URI from call locations, you'd need to read the URI from different positions in the hierarchy just to implement the reveal. WDYT?